pptx/opc/package.py

```diff
@@ -104,17 +108,15 @@ def main_document_part(self):
 
     def next_partname(self, tmpl):
         """
-        Return a |PackURI| instance representing the next available partname
+        Return a |PackURI| instance representing the next partname
         matching *tmpl*, which is a printf (%)-style template string
         containing a single replacement item, a '%d' to be used to insert the
         integer portion of the partname. Example: '/ppt/slides/slide%d.xml'
         """
-        partnames = [part.partname for part in self.iter_parts()]
-        for n in range(1, len(partnames) + 2):
-            candidate_partname = tmpl % n
-            if candidate_partname not in partnames:
-                return PackURI(candidate_partname)
-        raise Exception("ProgrammingError: ran out of candidate_partnames")
+        name = tmpl.split(os.sep)[2]
+        self.partnames[name] += 1
```

Comment:
I'm not seeing where `._partname` gets initialized with the part-names already present in the package. I think a better strategy would be to make `._partnames` an `@lazyproperty` (rather than an instance variable) and do the initialization in that lazyproperty (which only gets executed the first time it is called).